### ISSUE TITLE
Bug: Mask token is not tokenized correctly since v0.11.3

### ISSUE THREAD

**Issue by saattrupdan** [huggingface/tokenizers#918]
Since (Python) version 0.11.3, `tokenizers` are not dealing with mask tokens as single tokens, and instead split them up into characters.

## Minimal example
```python
from tokenizers import tokenizers, models, AddedToken, trainers

# Initialize tokenizer
model = models.BPE(unk_token='<unk>')
tokenizer = tokenizers.Tokenizer(model)

# Add special tokens
special_tokens = [
    AddedToken('<mask>', single_word=True, normalized=False),
    AddedToken('<unk>', single_word=True, normalized=False)
]
tokenizer.add_special_tokens(special_tokens)

# Add a white space pre-tokenizer
tokenizer.pre_tokenizer = pre_tokenizers.Whitespace()

# Initialise trainer
trainer = trainers.BpeTrainer(vocab_size=1000, special_tokens=special_tokens)

# Load corpus
corpus = ['Dette er en test']

# Train on dummy corpus
tokenizer.train_from_iterator(iterator=corpus, trainer=trainer)

# Test <mask> encoding
print(tokenizer.encode('Dette er <mask> test').tokens)
print(tokenizer.encode('<mask> test').tokens)
print(tokenizer.encode('<mask> ').tokens)
print(tokenizer.encode('Dette er <mask>').tokens)
print(tokenizer.encode('Dette er <mask> ').tokens)
print(tokenizer.encode('<mask>').tokens)
```

Correct output with `tokenizers==0.11.2`:
```
['Dette', 'er', '<mask>', 'test']
['<mask>', 'test']
['<mask>']
['Dette', 'er', '<mask>']
['Dette', 'er', '<mask>']
['<mask>']
```

Wrong output with `tokenizers` versions 0.11.3, 0.11.4 and the current latest version, 0.11.5:
```
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', 'test']
['<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', 'test']
['<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>']
['Dette', 'er', '<mask>']
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>']
['<mask>']
```

As can be seen in the above tests, the bug does not happen when the document ends with the mask token (but even a whitespace after is enough to trigger the bug).

**Comment by Narsil** [comment]
Thanks for the detailed bug report, with the easy to reproduce example.

I dived a bit into the code, and there seem that indeed an off-by one error was introduced. 
See the linked PR for the fix.

**Comment by saattrupdan** [comment]
@Narsil Thanks for your PR! That definitely solved my test cases in my previous post, but I just checked that it unfortunately still fails when the `<mask>` token is directly next to a non-space symbol. Here are some more test cases:
```python
test_docs = [
    'Dette er <mask>.',
    'Dette er <mask>,',
    'Dette er <mask>-',
    'Dette er <mask>0',
    'Dette er <mask>a',
]
for test_doc in test_docs:
    print(tokenizer.encode(test_doc).tokens)
```

For previous versions of `tokenizers` (with the training script above), this correctly outputs the following:
```python
['Dette', 'er', '<mask>', '<unk>']
['Dette', 'er', '<mask>', '<unk>']
['Dette', 'er', '<mask>', '<unk>']
['Dette', 'er', '<mask>', '<unk>']
['Dette', 'er', '<mask>', '<unk>']
```

However, with your PR, it outputs:
```python
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', '<unk>']
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', '<unk>']
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', '<unk>']
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', '<unk>']
['Dette', 'er', '<unk>', '<unk>', '<unk>', 's', '<unk>', '<unk>', '<unk>']
```

Your fix fixed all my previous cases though, so thanks a lot for that!

**Comment by Narsil** [comment]
Hi @saattrupdan ,

This is exactly what `single_word` option does. It prevents your special token from being parsed if it's not space separated.
If you want the symbol to be parsed no matter what, you need to use `single_word(False)`.

`single_word` option exists when you special token might be a subset of a regular word.
Imagining I am parsing old telegrams, `STOP`  might be a special token, but `STOPPING` might be a regular word in the text, so `single_word` exists to prevent from parsing as `STOP` + `PING`.

**Comment by saattrupdan** [comment]
@Narsil Aha, I see! So this was really a bug in the previous versions then, cf. my tests above?

**Comment by Narsil** [comment]
@saattrupdan Ignore what I said, the tests didn't reflect the full extent of what the thing was doing before, and the change that occurred between 0.11.2..0.11.3 (which was supposed to be a pure performance improvement) actually introduced a pretty BIG breaking change, moving away from regexp unicode definition, to only handle spaces (with a bug at that).

The PR has been updated to handle those things exactly as they were. The performance gains when using large amounts should still be there (since we're not handling a huge regexp like before) but could be somehow limited when using `lstrip`, `rstrip` and `single_word`. There is probably still room for performance improvement, but definitely not in the same range if we want to keep unicode correction in there.

Does this word better ? 

In your example, the first 3 examples should match, where the last 2 shouldn't (the last 0 is a "word" character, and "a" is one too)

**Comment by saattrupdan** [comment]
> @saattrupdan Ignore what I said, the tests didn't reflect the full extent of what the thing was doing before, and the change that occurred between 0.11.2..0.11.3 (which was supposed to be a pure performance improvement) actually introduced a pretty BIG breaking change, moving away from regexp unicode definition, to only handle spaces (with a bug at that).
> 
> The PR has been updated to handle those things exactly as they were. The performance gains when using large amounts should still be there (since we're not handling a huge regexp like before) but could be somehow limited when using `lstrip`, `rstrip` and `single_word`. There is probably still room for performance improvement, but definitely not in the same range if we want to keep unicode correction in there.
> 
> Does this word better ?
> 
> In your example, the first 3 examples should match, where the last 2 shouldn't (the last 0 is a "word" character, and "a" is one too)

@Narsil So, if I understand you correctly, when `single_word==True` then `'<mask>.'` _should_ be tokenized as `['<mask>', '.']`, but `'<mask>a'` should be tokenized as `['<mask>', 'a']`?

I ran my tests with the newest version of your branch, and that's indeed what's happening 🙂

**Comment by Narsil** [comment]
> should be tokenized as ['<mask>', 'a']
No as ['<mask>a'] (well, the model runs on it later so it could become a bunch of <unk> on the tiny examples we train)
